src/reducers/computeRenderedRows.js

```diff
@@ -239,3 +245,97 @@ function computeRenderedRowOffsets(state, rowRange) {
   state.rowOffsets = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+
+/**
+ * Walk the rows to render and compute the height offsets and
+ * positions in the row buffer.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdx: number,
+ *   endViewportIdx: number,
+ *   firstBufferIdx: number,
+ *   firstViewportIdx: number,
+ * }} rowRange
+ * @private
+ */
+function computeRenderedRowOffsetsInViewport(state, rowRange) {
+  const { rowBufferSet, rowOffsetIntervalTree, storedHeights } = state;
+  const {
+    endBufferIdx,
+    endViewportIdx,
+    firstBufferIdx,
+    firstViewportIdx,
+  } = rowRange;
+
+  const renderedRowsCount = endBufferIdx - firstBufferIdx;
+  if (renderedRowsCount === 0) {
+    state.rowOffsets = {};
+    state.rows = [];
+    return;
+  }
+
+  // output for this function
+  const rows = state.rows.slice(); // state.rows
+  const rowOffsets = {}; // state.rowOffsets
+
+  // incremental way for calculating rowOffset
+  let runningOffset = rowOffsetIntervalTree.sumUntil(firstViewportIdx);
+
+  // we add the rows inside the current view port and also calculate their offsets
+  for (let rowIdx = firstViewportIdx; rowIdx < endViewportIdx; rowIdx++) {
+    // Update the offset for rendering the row
+    rowOffsets[rowIdx] = runningOffset;
+    runningOffset += storedHeights[rowIdx];
+
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, rowBufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
+    rows[rowPosition] = rowIdx;
+  }
+
+  // we also need to store the rowOffsets for rows out side the view port that still exist in the buffer
+  const rowsInBufferSet = filter(state.rows, (rowIdx) => rowBufferSet.getValuePosition(rowIdx) !== null);
+  const rowsOutsideViewPort = filter(rowsInBufferSet, (rowIdx) => !inRange(rowIdx, firstViewportIdx, endViewportIdx));
+  rowsOutsideViewPort.forEach((rowIdx) => rowOffsets[rowIdx] = state.rowOffsets[rowIdx]);
```

Comment:
I think this comment from the previous review is not yet addressed:

> Could you walk me through how this ensures that a row from rowsNotInBuffer won't have an offset from the previous scroll which places into the buffer (or viewport)?
>
> The scenario I'm concerned about might look like:
> Row X is rendered at position 200, and I'm scrolled so 250 is the end of the buffer.
> I scroll up so 210 is at the end of the buffer, but in scrolling up, I find the rows newly added to the beginning of the buffer are larger than expected and Row X is actually positioned at 220.
> Could Row X be out of the buffer, but positioned in the viewport since it uses the old positioning of 200 instead of the new one of 220?

Reply:
Yeah, your concern is spot on.
We'll need to update the rowOffsets instead of using the old one. 
Unfortunately I couldn't figure out an efficient way to do this (mimicking the incremental way used for the view port rows), since the evicted rows are kind of random and can leave gaps.
Easiest solution is to just `rowOffsetIntervalTree.sumUntil` for the rows that aren't evicted but outside the viewport.

Reply:
This whole concern may go away if we address things the way I describe in my other comment above.